core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/ProgressiveUtils.scala

```diff
@@ -0,0 +1,38 @@
+package edu.uci.ics.texera.workflow.common
+
+import edu.uci.ics.texera.workflow.common.tuple.Tuple
+import edu.uci.ics.texera.workflow.common.tuple.schema.{Attribute, AttributeType}
+
+object ProgressiveUtils {
+
+  // boolean attribute to indicate insertion / retraction
+  // true  indicates insertion  (+)
+  // false indicates retraction (-)
+  val insertRetractFlagAttr = new Attribute("__internal_is_insertion", AttributeType.BOOLEAN)
+
+  def addInsertionFlag(tuple: Tuple): Tuple = {
+    assert(!tuple.getSchema.containsAttribute(insertRetractFlagAttr.getName))
+    Tuple.newBuilder.add(insertRetractFlagAttr, true).add(tuple).build
+  }
+
+  def addRetractionFlag(tuple: Tuple): Tuple = {
+    assert(!tuple.getSchema.containsAttribute(insertRetractFlagAttr.getName))
+    Tuple.newBuilder.add(insertRetractFlagAttr, false).add(tuple).build
+  }
+
+  def isInsertion(tuple: Tuple): Boolean = {
+    if (tuple.getSchema.containsAttribute(insertRetractFlagAttr.getName)) {
+      tuple.getField[Boolean](insertRetractFlagAttr.getName)
+    } else {
+      true
+    }
+  }
+
+  def getTupleFlagAndValue(tuple: Tuple): (Boolean, Tuple) = {
+    (
+      isInsertion(tuple),
+      Tuple.newBuilder().add(tuple).removeIfExists(insertRetractFlagAttr.getName).build()
```

Comment:
Any memory issue here?

Reply:
the same concern here. this is only probing the flag and value, will it require making a copy?

Reply:
This is out of scope of this PR